### PR TITLE
feat: C509 lib CLI example

### LINKED ISSUE #616: 🛠️ [TASK] : Create a cli for C509 library
### Summary

Create a cli for C509 library

### Description

Create a cli for C509 library

where
- TBS certificate in C509 can be generated
- C509 is generated from the TBS and private key using `Ed25519`
- given C509, verification on the signature is checked
- Deserialize the C509 back.

This ticket also include a `example` app for the library that implement a cli that can illustrate the above functionality

### PR DESCRIPTION
**Pull request by bkioshn**
# Description

Implement C509 lib CLI and example

## Related Issue(s)

> Closes #616 

### CLI
- `generate` command where it generate C509 certificate where it take a JSON file that contains necessary information to create the certificate. If the private key is provided, the certificate is signed.
- `verify` command where it verify the C509 signature
- `decode` command where it decode a C509 back to editable and readable file, in this case, JSON format is chosen. If the `output` is not given, the JSON will only be print out.
- Provide a sample JSON file use to generate C509

### Other
- Add PEM file handling in `signing.rs`
- Add additional necessary functions.

## Please confirm the following checks

* [x] My code follows the style guidelines of this project
* [x] I have performed a self-review of my code
* [x] I have commented my code, particularly in hard-to-understand areas
* [x] I have made corresponding changes to the documentation
* [x] My changes generate no new warnings
* [x] I have added tests that prove my fix is effective or that my feature works
* [x] New and existing unit tests pass locally with my changes
* [x] Any dependent changes have been merged and published in downstream module

### CODE REVIEW

File: catalyst-gateway-crates/c509-certificate/examples/cli/main.rs [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+use rand::Rng;
+use serde::{Deserialize, Serialize};
+
+/// A struct for CLI.
+#[derive(Parser)]
+#[command(version, about, long_about = None)]
+struct Cli {
+    /// Optional subcommands.
+    #[command(subcommand)]
+    command: Option<Commands>,
```
Comment (Mr-Leshiy) [inline catalyst-gateway-crates/c509-certificate/examples/cli/main.rs:34]: Seems it is redundant to have this `commands` field here.
Let's change this `Cli` struct to `enum` with copying basically all enum options from `Commands` and remove `Commands` type.
```rust
enum Cli {
Generate{...},
Verify{...},
....
}
```

File: catalyst-gateway-crates/c509-certificate/examples/cli/main.rs [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+    if self_signed && certificate_type.unwrap_or(SELF_SIGNED_INT) != SELF_SIGNED_INT {
+        return Err(anyhow::anyhow!(
+            "Certificate type should be 0 if self-signed is true"
+        ));
+    }
+    Ok(())
+}
+
+/// Parse public key from file path or string.
+fn parse_public_key(public_key: &str) -> anyhow::Result<PublicKey> {
```
Comment (Mr-Leshiy) [inline catalyst-gateway-crates/c509-certificate/examples/cli/main.rs:247]: I am not quite sure that this function will that way as you expected.
I would rather explicitly choose at the place or instantiating `PublicKey`  is it generated from file or from string.

**Review by Mr-Leshiy — CHANGES_REQUESTED** [review]

File: catalyst-gateway-crates/c509-certificate/examples/cli/main.rs [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+            Cli::Generate {
+                json_file,
+                output,
+                private_key,
+                key_type,
+            } => {
+                let sk = match private_key {
+                    Some(key) => Some(PrivateKey::from_file(key)?),
+                    None => None,
+                };
```
Comment (apskhem) [inline catalyst-gateway-crates/c509-certificate/examples/cli/main.rs:85]:
```suggestion
                let sk = private_key
                    .map(PrivateKey::from_file)
                    .transpose()?;
```

File: catalyst-gateway-crates/c509-certificate/Cargo.toml [hunk 4]
```diff
@@ -59,6 +59,17 @@ strum_macros = "0.26.3"
...
+serde = { version = "1.0.204", features = ["derive"] }
 
 [package.metadata.cargo-machete]
 ignored = ["strum"]
+
+[dev-dependencies]
+clap = { version = "4.5.9", features = ["derive"] }
+serde_json = "1.0.120"
+rand = "0.8.5"
+chrono = "0.4.38"
```
Comment (apskhem) [inline catalyst-gateway-crates/c509-certificate/Cargo.toml:71]: What are used or included during the runtime should move to `dependencies`. I think `clap` is the obvious one.

File: catalyst-gateway-crates/c509-certificate/examples/cli/main.rs [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+/// If self-signed is true, issuer is the same as subject.
+/// Otherwise, issuer should be present.
+fn determine_issuer(
+    self_signed: bool, issuer: Option<RelativeDistinguishedName>,
+    subject: RelativeDistinguishedName,
+) -> anyhow::Result<RelativeDistinguishedName> {
+    if self_signed {
+        Ok(subject)
+    } else {
+        issuer.ok_or_else(|| anyhow::anyhow!("Issuer should be present if self-signed is false"))
```
Comment (apskhem) [inline catalyst-gateway-crates/c509-certificate/examples/cli/main.rs:222]:
```suggestion
        issuer.ok_or_else(|| anyhow::anyhow!("Issuer must be present if self-signed is false"))
```

File: catalyst-gateway-crates/c509-certificate/examples/cli/main.rs [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+    }
+}
+
+/// Validate the certificate type.
+fn validate_certificate_type(
+    self_signed: bool, certificate_type: Option<u8>,
+) -> anyhow::Result<()> {
+    if self_signed && certificate_type.unwrap_or(SELF_SIGNED_INT) != SELF_SIGNED_INT {
+        return Err(anyhow::anyhow!(
+            "Certificate type should be 0 if self-signed is true"
```
Comment (apskhem) [inline catalyst-gateway-crates/c509-certificate/examples/cli/main.rs:232]:
```suggestion
            "Certificate type must be 0 if self-signed is true"
```

File: catalyst-gateway-crates/c509-certificate/examples/cli/main.rs (hunk 3 above)
Comment (bkioshn) [reply catalyst-gateway-crates/c509-certificate/examples/cli/main.rs:85]: I think this syntax is hard to interpret, will keep it this way, but thanks for the suggestion!

File: catalyst-gateway-crates/c509-certificate/Cargo.toml (hunk 4 above)
Comment (bkioshn) [reply catalyst-gateway-crates/c509-certificate/Cargo.toml:71]: Resolved since all of this is used in `examples`

**Review by Mr-Leshiy — APPROVED** [review]
LGTM

**Pull request merged**